src/report.js

```diff
@@ -0,0 +1,180 @@
+'use strict'
+
+const globals = require('./globals')
+const system = (process.platform === 'linux') ? require('./system.js') : require('./mockSystem.js')
+const os = require('os')
+const https = require('https')
+const log = console.log
+
+function sendRequest(requestBody, config) {
+  return new Promise((resolve, reject) => {
+    var req = https.request({
+      hostname: config.url,
+      path: config.path,
+      port: 443,
+      method: 'POST',
+      headers: {'content-type' : 'application/json'},
+      agent: globals.httpsAgent,
+      timeout: config.networkTimeout
+    }, (res) => {
+      var apiResponse = ''
+
+      res.on('data', function (chunk) {
+        apiResponse += chunk
+      })
+
+      res.on('end', function () {
+        resolve({ status: res.statusCode, apiResponse })
+      })
+    }).on('error', (err) => {
+      reject(err)
+    })
+
+    req.write(JSON.stringify(requestBody))
+    req.end()
+  })
+}
+
+class Report {
+  constructor(config, context, startTime, metrics) {
+    this.initalPromises = {
+      statPromise: system.readstat('self'),
+      bootIdPromise: system.readbootid()
+    }
+
+    // flag on report sending status, reports are sent once
+    this.sent = false
+
+    this.config = config
+    this.context = context
+    this.startTime = startTime
+
+    // Populate initial report skeleton on construction
+    this.report = {
+      client_id: config.clientId,
+      installMethod: config.installMethod,
+      duration: undefined,
+      processId: globals.PROCESS_ID,
+      aws: {
+        functionName: context.functionName,
+        functionVersion: context.functionVersion,
+        awsRequestId: context.awsRequestId,
+        invokedFunctionArn: context.invokedFunctionArn,
+        logGroupName: context.logGroupName,
+        logStreamName: context.logStreamName,
+        memoryLimitInMB: context.memoryLimitInMB,
+        getRemainingTimeInMillis: undefined,
+        traceId: process.env._X_AMZN_TRACE_ID
+      },
+      environment: {
+        agent: {
+          runtime: 'nodejs',
+          version: globals.VERSION,
+          load_time: globals.MODULE_LOAD_TIME
+        },
+        nodejs: {
+          version: process.version,
+          memoryUsage: undefined
+        },
+        host: {
+          container_id: undefined
+        },
+        os: {},
+      },
+      errors: {},
+      coldstart: globals.COLDSTART,
+      custom_metrics: metrics,
+    }
+  }
+
+  send(err, callback) {
+
+    // Send report only once
+    if (this.sent) {
+      return
+    }
+    const self = this
+    const config = this.config
+    const context = this.context
+
+    // Add error to report if necessary
+    if (err) {
```

Comment:
a user asked if we could expose an err function in Node as we do in Python. Maybe a separate PR though...